src/go/rpk/pkg/cli/cluster/partitions/list.go

```diff
@@ -80,26 +89,39 @@ List all in json format.
 			var mu sync.Mutex
 			if len(topics) == 0 && all {
 				clusterPartitions, err = cl.AllClusterPartitions(cmd.Context(), true, disabledOnly)
+				// If the admin API returns a 404, most likely rpk is talking
+				// with an old cluster.
+				if he := (*adminapi.HTTPResponseError)(nil); errors.As(err, &he) {
+					if he.Response.StatusCode == http.StatusNotFound {
+						out.Die("unable to query all partitions in the cluster: %v You may need to upgrade the cluster to access this feature or try listing per-topic with: 'rpk cluster partition list [TOPICS]'", err)
+					}
+				}
 				out.MaybeDie(err, "unable to query all partitions in the cluster: %v", err)
 			} else {
 				g, egCtx := errgroup.WithContext(cmd.Context())
 				for _, topic := range topics {
 					if topic == "" {
 						out.Die("invalid empty topic\n%v", cmd.UsageString())
 					}
-					nsTopic := strings.SplitN(topic, "/", 2)
-					var ns, topicName string
-					if len(nsTopic) == 1 {
-						ns = "kafka"
-						topicName = nsTopic[0]
-					} else {
-						ns = nsTopic[0]
-						topicName = nsTopic[1]
-					}
+					ns, topicName := nsTopic(topic)
 					g.Go(func() error {
 						cPartition, err := cl.TopicClusterPartitions(egCtx, ns, topicName, disabledOnly)
 						if err != nil {
-							return fmt.Errorf("unable to query cluster partition metadata of topic %q: %v", topicName, err)
+							// If the admin API returns a 404, most likely rpk
+							// is talking with an old cluster.
+							var he *adminapi.HTTPResponseError
+							isNotFoundErr := errors.As(err, &he) && he.Response.StatusCode == http.StatusNotFound
+							if !isNotFoundErr {
+								return fmt.Errorf("unable to query cluster partition metadata of topic %q: %v", topicName, err)
+							}
+							zap.L().Sugar().Warn("Admin API 'GET /v1/cluster/partitions' returned 404, trying now with 'GET /v1/partitions'")
+							if disabledOnly {
+								zap.L().Sugar().Warn("--only-disabled flag is not supported in this API version")
+							}
+							cPartition, err = topicPartitions(egCtx, cl, ns, topicName)
```

Comment:
Why did #14025 need to also issue a request per-partition? It looks like all the info is available in just the topic req?

Reply:
Yes, it's available on the topic req, I guess per-partition is to avoid requesting large topic metadata req, and instead just request the partition info you are interested in.

However, I went with the "issue for the whole topic -> then filter" because I felt like adding this optimization here is more convoluted for only a fallback mechanism that will be fairly unused in a couple of months.